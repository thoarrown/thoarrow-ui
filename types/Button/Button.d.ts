/// <reference types="react" />
import { ButtonProps } from "./Button.types";
declare const Button: ({ className, onClick, primary, size, text, appearance, children, disabled, iconOnly, ...rest }: ButtonProps) => JSX.Element;
export default Button;
