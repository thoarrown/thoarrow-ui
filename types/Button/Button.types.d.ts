import { MouseEventHandler, ButtonHTMLAttributes } from "react";
export declare type ButtonSize = "default" | "small" | "large";
export declare type ButtonAppearance = "default" | "secondary" | "danger";
export declare type ButtonProps = ButtonHTMLAttributes<unknown> & {
    text?: string;
    primary?: boolean;
    disabled?: boolean;
    iconOnly?: boolean;
    size?: ButtonSize;
    appearance?: ButtonAppearance;
    onClick?: MouseEventHandler<HTMLButtonElement>;
};
