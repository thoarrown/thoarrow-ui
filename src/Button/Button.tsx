import React from "react";
import styled, {
  css,
  DefaultTheme,
  FlattenInterpolation,
  ThemeProps,
} from "styled-components";
import { ButtonAppearance, ButtonProps, ButtonSize } from "./Button.types";

type CssPartial = FlattenInterpolation<ThemeProps<DefaultTheme>>;

const StyledButton = styled.button<{
  size: CssPartial;
  colors: CssPartial;
  disabled: boolean;
  iconOnly: boolean;
}>`
  border: 1px solid transparent;
  font: inherit;
  ${(props) => props.size};
  ${(props) => props.colors};
`;

const Button = ({
  className = "",
  onClick,
  size = "default",
  text,
  appearance = "default",
  children,
  disabled = false,
  iconOnly = false,
  icon,
  iconRight,
  ...rest
}: ButtonProps) => {
  return (
    <StyledButton
      className={className}
      size={getButtonSize(size)}
      colors={getColors(appearance, disabled)}
      disabled={disabled}
      iconOnly={iconOnly}
      {...rest}
    >
      {icon && <span className="button-icon">{icon}</span>}
      {children && children}
      {iconRight && (
        <span className="button-icon button-icon-right">{iconRight}</span>
      )}
    </StyledButton>
  );
};

function getColors(appearance: ButtonAppearance, disabled: boolean) {
  if (disabled) {
    return css`
      background-color: ${(props) => props.theme.colors.disabled};
    `;
  }

  if (appearance === "danger") {
    return css`
      background-color: ${(props) => props.theme.colors.warning};
    `;
  }

  if (appearance === "secondary") {
    return css`
      background-color: ${(props) => props.theme.colors.secondary};
    `;
  }

  return css`
    background-color: ${(props) => props.theme.colors.primary};
  `;
}

function getButtonSize(size: ButtonSize) {
  if (size === "small") {
    return css`
      height: 2.5rem;
    `;
  }
  return css`
    height: 3.5rem;
  `;
}
export default Button;
