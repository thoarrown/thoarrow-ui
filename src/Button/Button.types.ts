import { ButtonHTMLAttributes, ReactNode } from "react";

export type ButtonSize = "default" | "small" | "large";
export type ButtonAppearance = "default" | "secondary" | "danger";

export type ButtonProps = ButtonHTMLAttributes<unknown> & {
  text?: string;
  disabled?: boolean;
  iconOnly?: boolean;
  size?: ButtonSize;
  appearance?: ButtonAppearance;
  icon?: ReactNode;
  iconRight?: ReactNode;
};
