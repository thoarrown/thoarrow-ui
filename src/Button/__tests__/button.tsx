import React from 'react';
import { render } from '../../../tests/utils';
import Button from '../Button';

describe('Button', () => {
  it('should support all sizes', () => {
    const { container } = render(
      <>
        <Button size="default" />
        <Button size="large" />
      </>,
    );
    expect(container).toMatchSnapshot();
  });
});
