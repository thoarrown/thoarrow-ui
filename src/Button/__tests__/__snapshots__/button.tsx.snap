// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should support all sizes 1`] = `
.c0 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  background-color: #6C5DD3;
}

<div>
  <button
    class="c0"
  />
  <button
    class="c0"
  />
</div>
`;
